code/modules/modular_computers/file_system/programs/ntdownloader.dm

```diff
@@ -146,48 +154,62 @@
 	var/obj/item/computer_hardware/hard_drive/hard_drive = my_computer.all_components[MC_HDD]
 	data["disk_size"] = hard_drive.max_capacity
 	data["disk_used"] = hard_drive.used_capacity
-	var/list/all_entries[0]
+	var/list/programs[0]
+	var/list/programs_incompatible[0]
 	for(var/A in main_repo)
 		var/datum/computer_file/program/P = A
-		// Only those programs our user can run will show in the list
-		if(hard_drive.find_file_by_name(P.filename))
-			continue
-		all_entries.Add(list(list(
-			"filename" = P.filename,
-			"filedesc" = P.filedesc,
-			"fileinfo" = P.extended_desc,
-			"compatibility" = check_compatibility(P),
-			"size" = P.size,
-			"access" = P.can_run(user,transfer = 1, access = access)
-		)))
-	data["hackedavailable"] = FALSE
+		if (check_compatibility(P))
+			programs.Add(P)
+		else
+			programs_incompatible.Add(P)
 	if(emagged) // If we are running on emagged computer we have access to some "bonus" software
-		var/list/hacked_programs[0]
 		for(var/S in antag_repo)
 			var/datum/computer_file/program/P = S
-			if(hard_drive.find_file_by_name(P.filename))
-				continue
-			data["hackedavailable"] = TRUE
-			hacked_programs.Add(list(list(
+			if (check_compatibility(P))
+				programs.Add(P)
+			else
+				programs_incompatible.Add(P)
+
+	programs = sortList(programs, /proc/cmp_program_asc)
+	programs_incompatible = sortList(programs_incompatible, /proc/cmp_program_asc)
+
+	var/categories = show_categories.Copy()
+	for(var/V in categories)
+		categories[V] = list()
+	for(var/I in programs | programs_incompatible)
+		var/datum/computer_file/program/P = I
+		for(var/C in categories)
+			if(C == P.category || C == PROGRAM_CATEGORY_ALL)
+				categories[C] += P
+	for(var/category in categories)
+		if (!LAZYLEN(categories[category]))
+			continue
+		var/list/cat = list(
+			"name" = category,
+			"items" = list())
+		for(var/I in categories[category])
+			var/datum/computer_file/program/P = I
+			cat["items"] += list(list(
+				"icon" = P.program_icon,
 				"filename" = P.filename,
 				"filedesc" = P.filedesc,
 				"fileinfo" = P.extended_desc,
+				"category" = P.category,
```

Comment:
You already have category property here, there is no need to have nested lists by category, i.e. `list(category = list(...))`